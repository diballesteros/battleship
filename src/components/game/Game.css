.game-page {
    height: 100%;
    width: 100%;
}

.game-view {
    height: 100%;
    display: flex;
    flex-flow: column nowrap;
    align-items: center;
    justify-content: center;
}

.game-title {
    font-size: 64px;
    font-weight: bold;
    color: #FFFFFF;
    margin-top: 15px;
    margin-bottom: 0px;
}