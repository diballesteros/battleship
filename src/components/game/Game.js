import React, { Component } from 'react';
import PlayerView from '../playerview/PlayerView';

import _ from 'lodash';
import ShipFactory from '../shipfactory/ShipFactory'

const beginningSquares = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90]
const endSquares = [9, 19, 29, 39, 49, 59, 69, 79, 89, 99];

class Game extends Component {
    constructor(props) {
        super(props);
        this.state = {
            computerMoves: [...Array(100).keys()],
            lastSuccessfulMoves: [],
            successfulComputerHit: false
        }
    };

    removeComputerMove(i) {
        const modifiedComputerMoves = this.state.computerMoves.slice();
        const computerHitIndex = this.state.computerMoves.findIndex(move => move === i);
        modifiedComputerMoves.splice(computerHitIndex, 1);
        this.setState({
            computerMoves: modifiedComputerMoves
        })
    };

    getComputerMove() {
        let nextMove;
        if (this.state.successfulComputerHit === true) {
            const startingPoint = this.state.lastSuccessfulMoves.length;
            nextMove = this.getAdjacentSquare(startingPoint - 1);
        } 
        if (_.isUndefined(nextMove) || _.isNull(nextMove)) {
            nextMove = _.sample(this.state.computerMoves);
        }
        this.removeComputerMove(nextMove);
        return nextMove;
    };

    getAdjacentSquare() {
        let square = 0;
        if (this.state.lastSuccessfulMoves.length >= 2) {
            if ((this.state.lastSuccessfulMoves[0] + 1 === this.state.lastSuccessfulMoves[1]) || (this.state.lastSuccessfulMoves[0] - 1 === this.state.lastSuccessfulMoves[1])) {
                for (let i = this.state.lastSuccessfulMoves.length - 1; i > -1; i--) {
                    square = this.getAdjacentHorizontalSquare(i);
                    if (_.isNumber(square)) {
                        return square;
                    }
                }
            } else {
                for (let i = this.state.lastSuccessfulMoves.length - 1; i > -1; i--) {
                    square = this.getAdjacentVerticalSquare(i);
                    if (_.isNumber(square)) {
                        return square;
                    }
                }
            }
        } else {
            for (let i = this.state.lastSuccessfulMoves.length - 1; i > -1; i--) {
                square = this.getAdjacentHorizontalSquare(i);
                if (_.isNumber(square)) {
                    return square;
                }
                square = this.getAdjacentVerticalSquare(i);
                if (_.isNumber(square)) {
                    return square;
                }
            }
        }
    };

    getAdjacentHorizontalSquare(i) {
        if (this.state.computerMoves.includes(this.state.lastSuccessfulMoves[i] - 1) && !beginningSquares.includes(this.state.lastSuccessfulMoves[i])) {
            return (this.state.lastSuccessfulMoves[i] - 1);
        } else if (this.state.computerMoves.includes(this.state.lastSuccessfulMoves[i] + 1) && !endSquares.includes(this.state.lastSuccessfulMoves[i])) {
            return (this.state.lastSuccessfulMoves[i] + 1);
        }
    }
    getAdjacentVerticalSquare(i) {
        if (this.state.computerMoves.includes(this.state.lastSuccessfulMoves[i] + 10)) {
            return (this.state.lastSuccessfulMoves[i] + 10);
        } else if (this.state.computerMoves.includes(this.state.lastSuccessfulMoves[i] - 10)) {
            return (this.state.lastSuccessfulMoves[i] - 10);
        }
    }

    setSuccessfulHit(move) {
        const modifiedSuccessfulMoves = this.state.lastSuccessfulMoves.slice();
        modifiedSuccessfulMoves.push(move);
        this.setState({
            successfulComputerHit: true,
            lastSuccessfulMoves: modifiedSuccessfulMoves
        });
    };

    resetSuccessfulHit() {
        this.setState({
            successfulComputerHit: false,
            lastSuccessfulMoves: []
        });
    }

    render() {
        return (
            <div>
                <ShipFactory />
                <PlayerView
                    getComputerMove={() => this.getComputerMove()}
                    setSuccessfulHit={(move) => this.setSuccessfulHit(move)}
                    resetSuccessfulHit={(() => this.resetSuccessfulHit())} />
            </div>
        );
    };
};

export default Game;
