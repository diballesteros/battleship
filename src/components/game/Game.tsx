/* eslint-disable no-plusplus */
import React, { useReducer } from 'react';
import { toast } from 'react-toastify';
import _ from 'lodash';
import { GameState, GameActions, Ship } from 'constants/Types';
import { INITIAL_STATE, FIRSTCOLUMNSQUARES, LASTCOLUMNSQUARES } from 'constants/constant';
import Factory from 'containers/Factory/Factory';
import NewGameBoard from 'components/NewGameBoard/NewGameBoard';
import PlayerView from '../playerview/PlayerView';
import ShipFactory from '../shipfactory/ShipFactory';
import Modal from '../UI/Modal/Modal';
import styles from './Game.module.scss';

const gameReducer = (state: GameState, action: GameActions) => {
  switch (action.type) {
    case 'SET_BOARD':
      return { ...INITIAL_STATE, ...action.payload };
    case 'PLAYER_MOVE':
      return { ...state, ...action.payload };
    case 'COMPUTER_MOVE':
      return { ...state, ...action.payload };
    case 'UPDATE_HIT':
      return { ...state, ...action.payload };
    case 'RESET_BOARD':
      return { ...INITIAL_STATE };
    default:
      return state;
  }
};

const Game: React.FC = () => {
  const [state, dispatch] = useReducer(gameReducer, INITIAL_STATE);

  const setShips = (builtShips: any, builtComputerShips: any) => {
    dispatch({
      type: 'SET_BOARD',
      payload: { playerShips: builtShips, computerShips: builtComputerShips },
    });
  };

  const removeComputerMove = (i: number) => {
    const modifiedComputerMoves = state.computerMoves.slice();
    const computerHitIndex = state.computerMoves.findIndex((move: number) => move === i);
    modifiedComputerMoves.splice(computerHitIndex, 1);
    dispatch({ type: 'COMPUTER_MOVE', payload: { computerMoves: modifiedComputerMoves } });
  };

  const getAdjacentHorizontalSquare = (i: number) => {
    if (
      state.computerMoves.includes(state.lastSuccessfulMoves[i] - 1) &&
      !FIRSTCOLUMNSQUARES.includes(state.lastSuccessfulMoves[i])
    ) {
      return state.lastSuccessfulMoves[i] - 1;
    }
    if (
      state.computerMoves.includes(state.lastSuccessfulMoves[i] + 1) &&
      !LASTCOLUMNSQUARES.includes(state.lastSuccessfulMoves[i])
    ) {
      return state.lastSuccessfulMoves[i] + 1;
    }
    return null;
  };

  const getAdjacentVerticalSquare = (i: number) => {
    if (state.computerMoves.includes(state.lastSuccessfulMoves[i] + 10)) {
      return state.lastSuccessfulMoves[i] + 10;
    }
    if (state.computerMoves.includes(state.lastSuccessfulMoves[i] - 10)) {
      return state.lastSuccessfulMoves[i] - 10;
    }
    return null;
  };

  const getAdjacentSquare = () => {
    let square = 0;
    if (state.lastSuccessfulMoves.length >= 2) {
      if (
        state.lastSuccessfulMoves[0] + 1 === state.lastSuccessfulMoves[1] ||
        state.lastSuccessfulMoves[0] - 1 === state.lastSuccessfulMoves[1]
      ) {
        for (let i = state.lastSuccessfulMoves.length - 1; i > -1; i--) {
          square = getAdjacentHorizontalSquare(i);
          if (_.isNumber(square)) {
            return square;
          }
        }
      } else {
        for (let i = state.lastSuccessfulMoves.length - 1; i > -1; i--) {
          square = getAdjacentVerticalSquare(i);
          if (_.isNumber(square)) {
            return square;
          }
        }
      }
    } else {
      for (let i = state.lastSuccessfulMoves.length - 1; i > -1; i--) {
        square = getAdjacentHorizontalSquare(i);
        if (_.isNumber(square)) {
          return square;
        }
        square = getAdjacentVerticalSquare(i);
        if (_.isNumber(square)) {
          return square;
        }
      }
    }
    return null;
  };

  const getComputerMove = () => {
    let nextMove;
    if (state.successfulComputerHit === true) {
      nextMove = getAdjacentSquare();
    }
    if (_.isUndefined(nextMove) || _.isNull(nextMove)) {
      nextMove = _.sample(state.computerMoves);
    }
    removeComputerMove(nextMove);
    return nextMove;
  };

  const setSuccessfulHit = (move: number) => {
    const modifiedSuccessfulMoves = state.lastSuccessfulMoves.slice();
    modifiedSuccessfulMoves.push(move);
    dispatch({
      type: 'UPDATE_HIT',
      payload: { lastSuccesfulMoves: modifiedSuccessfulMoves, successfulComputerHit: true },
    });
  };

  const resetSuccessfulHit = () => {
    dispatch({
      type: 'UPDATE_HIT',
      payload: { lastSuccesfulMoves: [], successfulComputerHit: false },
    });
  };

  const isSunk = (positions: any, isComputer: boolean) => {
    const sunkStatus = positions.every((position: boolean) => position === true);
    if (sunkStatus) {
      toast.info(
        `${isComputer ? 'Allied battleship has been sunk!' : 'Enemy battleship has been sunk!'}`,
        {
          position: toast.POSITION.BOTTOM_CENTER,
        }
      );
    }
    if (sunkStatus && isComputer === true) {
      resetSuccessfulHit();
    }
    return sunkStatus;
  };

  const resolveBoardState = (move: number, shipId: number, ships: any, isComputer: boolean) => {
    const shipIndex = ships.findIndex((ship: any) => ship.id === shipId);
    const hitIndex = ships[shipIndex].positions.indexOf(move);
    ships[shipIndex].hits[hitIndex] = true;
    ships[shipIndex].isSunk = isSunk(ships[shipIndex].hits, isComputer);
    return ships;
  };

  const resolveComputerTurn = (move: number) => {
    let modifiedPlayerShips = state.playerShips.slice();
    let foundShip = -1;
    modifiedPlayerShips.forEach((ship) => {
      if (ship.positions.includes(move)) {
        foundShip = ship.id;
      }
    });
    if (foundShip > -1) {
      setSuccessfulHit(move);
      modifiedPlayerShips = resolveBoardState(move, foundShip, modifiedPlayerShips, true);
    }
    return modifiedPlayerShips;
  };

  const resolveGameState = (playerShips: any, computerShips: any) => {
    const playerShipsSunk = _.every(playerShips, (ship) => ship.isSunk === true);
    const computerShipsSunk = _.every(computerShips, (ship) => ship.isSunk === true);
    if (playerShipsSunk) {
      toast.error(`${'You lose!'}`, {
        position: toast.POSITION.BOTTOM_CENTER,
      });
      return true;
    }
    if (computerShipsSunk) {
      toast.success(`${'You win!'}`, {
        position: toast.POSITION.BOTTOM_CENTER,
      });
      return true;
    }
    return false;
  };

  const receivePlayerAttack = (shipId: number, playerMove: number) => {
    let modifiedComputerShips = state.computerShips.slice();
    if (!_.isNull(shipId)) {
      modifiedComputerShips = resolveBoardState(
        playerMove,
        shipId,
        state.computerShips.slice(),
        false
      );
    }

    const computerMove = getComputerMove();
    const modifiedPlayerShips = resolveComputerTurn(computerMove);

    const gameResolved = resolveGameState(modifiedPlayerShips, modifiedComputerShips);

    dispatch({
      type: 'PLAYER_MOVE',
      payload: {
        computerShips: modifiedComputerShips,
        playerMoves: [...state.playerMoves, playerMove],
        gameResolved,
        playerShips: modifiedPlayerShips,
        completedComputerMoves: [...state.completedComputerMoves, computerMove],
      },
    });
  };

  const resetGame = () => {
    dispatch({ type: 'RESET_BOARD' });
  };

  return (
    <div className={styles.game_page}>
      <h1 className={styles.game_title}>BATTLESHIP</h1>
      <div className={styles.areaOne}>
        <NewGameBoard />
      </div>
      <div className={styles.areaTwo}>
        <Factory />
      </div>
    </div>
  );
};

export default Game;
