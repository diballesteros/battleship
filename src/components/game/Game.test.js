import React from 'react';
import { render } from '@testing-library/react';
import Game from './Game';

describe('Game component', () => {
    test('Initialized board', () => {
        const { asFragment } = render(<Game />);
        expect(asFragment()).toMatchSnapshot();
    });
});
