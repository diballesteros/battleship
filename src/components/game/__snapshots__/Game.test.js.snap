// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component Initialized board 1`] = `
<DocumentFragment>
  <div
    class="game-page"
  >
    <div
      class="game-view"
    >
      <label
        class="game-title"
      >
        BATTLESHIP
      </label>
      <div
        class="ship_factory"
      >
        <div>
          <div
            class="letter_row"
          >
            <div
              class="square grid-square"
            >
               
            </div>
            <div
              class="square grid-square"
            >
              A
            </div>
            <div
              class="square grid-square"
            >
              B
            </div>
            <div
              class="square grid-square"
            >
              C
            </div>
            <div
              class="square grid-square"
            >
              D
            </div>
            <div
              class="square grid-square"
            >
              E
            </div>
            <div
              class="square grid-square"
            >
              F
            </div>
            <div
              class="square grid-square"
            >
              G
            </div>
            <div
              class="square grid-square"
            >
              H
            </div>
            <div
              class="square grid-square"
            >
              I
            </div>
            <div
              class="square grid-square"
            >
              J
            </div>
          </div>
          <div
            class="lower_section"
          >
            <div
              class="number_column"
            >
              <div
                class="square grid-square"
              >
                0
              </div>
              <div
                class="square grid-square"
              >
                1
              </div>
              <div
                class="square grid-square"
              >
                2
              </div>
              <div
                class="square grid-square"
              >
                3
              </div>
              <div
                class="square grid-square"
              >
                4
              </div>
              <div
                class="square grid-square"
              >
                5
              </div>
              <div
                class="square grid-square"
              >
                6
              </div>
              <div
                class="square grid-square"
              >
                7
              </div>
              <div
                class="square grid-square"
              >
                8
              </div>
              <div
                class="square grid-square"
              >
                9
              </div>
            </div>
            <div
              class="board"
            >
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
              <div
                class="square ocean-square my_square"
              />
            </div>
          </div>
        </div>
        <div
          class="ship_store"
        >
          <div
            class="ship_store_title"
          >
            <label>
              Arrange your board
            </label>
            <span>
              Drag and drop the ship below on a square
            </span>
          </div>
          <div
            class="ship_store_ship_container"
          >
            <div
              class="ship_store_model"
            >
              <label>
                Model: Carrier
              </label>
            </div>
            <div
              class="ship_builder"
            >
              <div
                class="ship_store_ship horizontal"
                draggable="true"
              >
                <div
                  class="square ship-square my_square"
                />
                <div
                  class="square ship-square my_square"
                />
                <div
                  class="square ship-square my_square"
                />
                <div
                  class="square ship-square my_square"
                />
                <div
                  class="square ship-square my_square"
                />
              </div>
            </div>
            <div
              class="ship_store_counter"
            >
              Ships left: 5/5
            </div>
            <div
              class="ship_store_buttons"
            >
              <button
                disabled=""
              >
                Start Game
              </button>
              <button>
                Rotate
              </button>
              <button
                disabled=""
              >
                Undo
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
